File: Flywaight/cpp/FabricaDeLineas.cpp
#include <bits/stdc++.h>
#include "LineaLigera.cpp"

using namespace std;

class FabricaDeLineas{
    private:
        list<IlineaLigera*> lineas;
        list<IlineaLigera*>::iterator i;
    public:
        FabricaDeLineas();
        IlineaLigera* getLine( string color );

};

FabricaDeLineas::FabricaDeLineas(){
    lineas.clear();
    i=lineas.begin();
}

IlineaLigera* FabricaDeLineas::getLine( string color ){
    // Comprobar si hemos creado una línea con el color solicitado, y devolverla en tal caso
     for(i=lineas.begin();i!=lineas.end();i++){
        IlineaLigera* a=*i;
        if(a->getColor()==color){
            cout<<"Línea de color [" + color + "] encontrada, la devolvemos"<<endl;
            return a;
        }
    }
    // Si no ha sido creada la creamos ahora, la agregamos a la lista y la devolvemos
    cout<<"Creando una línea de color [" + color + "]"<<endl;
    IlineaLigera* linea = new Linea( color );
     this->lineas.push_front( linea );
    return linea;
}    

File: Flywaight/Java/Fabrica/FabricaDeLineas.java
package Fabrica;

import java.util.ArrayList;
import java.util.List;
import LineaLigera.*;
import ConcretLineal.*;

public class FabricaDeLineas {

    private List<ILineaLigera> lineas;
    // --------------------
    public FabricaDeLineas(){
         this.lineas = new ArrayList<ILineaLigera>();
    }
    // --------------------
     public ILineaLigera getLine( String color ){
        // Comprobar si hemos creado una línea con el color solicitado, y devolverla en tal caso
         for(ILineaLigera linea : this.lineas){
             if( linea.getColor().equals(color) ){
                System.out.println("Línea de color [" + color + "] encontrada, la devolvemos");
                return linea;
            }
        }
        // Si no ha sido creada la creamos ahora, la agregamos a la lista y la devolvemos
        System.out.println("Creando una línea de color [" + color + "]");
         ILineaLigera linea = new Lineal( color );
         this.lineas.add( linea );
        return linea;
    }
}
